### PR TITLE
Recompute Offload v1

### PR DESCRIPTION
**Pull request by JZ-LIANG**
### PR types
New features

### PR changes
OPs

### Describe
Recompute Offload version 1： memcpy & calc in same stream

### CODE REVIEW

File: paddle/fluid/operators/assign_op.cc [hunk 1]
```diff
@@ -107,10 +107,11 @@ class AssignKernel {
         ctx.HasOutput("Out"), true,
         platform::errors::NotFound("Output(Out) of assign_op is not found."));
     auto *out = ctx.OutputVar("Out");
+    auto kind = ctx.Attr<int>("kind");
```
Comment (gongweibao) [inline paddle/fluid/operators/assign_op.cc]: `kind` 这个名字过于 general了，用`place_type`?

File: paddle/fluid/operators/assign_op.h [hunk 2]
```diff
@@ -76,6 +90,7 @@ class AssignFunctor {
 
   framework::Variable *out_;
   const platform::DeviceContext &dev_ctx_;
+  const int kind_;
```
Comment (gongweibao) [inline paddle/fluid/operators/assign_op.h]: Set default value?

File: paddle/fluid/operators/fill_constant_op.cc [hunk 3]
```diff
@@ -116,6 +116,11 @@ class FillConstantOpMaker : public framework::OpProtoAndCheckerMaker {
                   "memory. Otherwise, fill output variable to the running "
                   "device")
         .SetDefault(false);
+    AddAttr<bool>(
+        "pinned",
```
Comment (gongweibao) [inline paddle/fluid/operators/fill_constant_op.cc]: `place_type`?

File: python/paddle/fluid/optimizer.py [hunk 4]
```diff
@@ -4703,6 +4708,389 @@ def mlp(input_x, input_y, hid_dim=128, label_dim=2):
...
+                                input_var))
+
+        assert len(self.un_fetch_checkpoint_names
+                   ) == 0, "{} checkpoints have NOT been Recorded".format(
+                       self.un_fetch_checkpoint_names)
+
+    def _update_backword(self):
+        if len(self.idx2insertions) == 0:
+            return
+        total_op = len(self.block.ops)
```
Comment (gongweibao) [inline python/paddle/fluid/optimizer.py:4895]: Add more comments?

File: python/paddle/fluid/optimizer.py [hunk 5]
```diff
@@ -4703,6 +4708,389 @@ def mlp(input_x, input_y, hid_dim=128, label_dim=2):
...
+        for idx, op in enumerate(self.block.ops):
+            if int(op.desc.attr("op_role")) == 0:
+                self.fw_strart_op_idx = idx
+                break
+
+        assert self.fw_strart_op_idx < len(
+            self.block.ops), "Could NOT found Forward op in prog"
+        last_offload_checkpoint = None
+
+        for i, op in enumerate(self.block.ops[self.fw_strart_op_idx:
```
Comment (gongweibao) [inline python/paddle/fluid/optimizer.py:4938]: This function is too long, split it?

File: python/paddle/fluid/optimizer.py [hunk 6]
```diff
@@ -4703,6 +4708,389 @@ def mlp(input_x, input_y, hid_dim=128, label_dim=2):
...
+            self.checkpoint_name2fetch_name = dict()
+            # self.checkpoint_name2load_name = dict()
+            for checkpoint_varname in self.sorted_checkpoint_names:
+                pinned_var_name, fetch_var_name = self.creat_vars(
+                    checkpoint_varname)
+                self.checkpoint_name2pinned_name[
+                    checkpoint_varname] = pinned_var_name
+                self.checkpoint_name2fetch_name[
+                    checkpoint_varname] = fetch_var_name
+                # self.checkpoint_name2load_name[checkpoint_var.name] = offload_var_name
```
Comment (gongweibao) [inline python/paddle/fluid/optimizer.py]: Clean up this?

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/assign_op.cc (hunk 1 above)
Comment (JZ-LIANG) [reply paddle/fluid/operators/assign_op.cc]: done

File: paddle/fluid/operators/fill_constant_op.cc (hunk 3 above)
Comment (JZ-LIANG) [reply paddle/fluid/operators/fill_constant_op.cc]: done

File: python/paddle/fluid/optimizer.py (hunk 4 above)
Comment (JZ-LIANG) [reply python/paddle/fluid/optimizer.py:4895]: done

**Pull request closed without merging**